File: backend/utils/geo_utils.py
import math
from scipy.spatial import KDTree

def haversine(a, b):
    lon1, lat1 = a
    lon2, lat2 = b
    R = 6371
    dlon, dlat = math.radians(lon2 - lon1), math.radians(lat2 - lat1)
    h = math.sin(dlat/2)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon/2)**2
    return 2 * R * math.asin(math.sqrt(h))

class NodeLocator:
    def __init__(self, nodes):
        self.nodes = nodes
        self.tree = KDTree(nodes)

    def find_nearest(self, coord):
        _, idx = self.tree.query(coord)
        return self.nodes[idx]

File: backend_java/MapProject/src/main/java/com/example/pathfinding/util/GeoUtils.java
package com.example.pathfinding.util;

import com.example.pathfinding.model.Node;

public class GeoUtils {

    public static double haversine(Node a, Node b) {
        double lon1 = a.getLon();
        double lat1 = a.getLat();
        double lon2 = b.getLon();
        double lat2 = b.getLat();

        double R = 6371.0; // km
        double dLon = Math.toRadians(lon2 - lon1);
        double dLat = Math.toRadians(lat2 - lat1);

        double h = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                Math.cos(Math.toRadians(lat1)) *
                        Math.cos(Math.toRadians(lat2)) *
                        Math.sin(dLon / 2) * Math.sin(dLon / 2);

        return 2 * R * Math.asin(Math.sqrt(h));
    }
}
